dev-tools/mage/common.go

```diff
@@ -532,42 +525,41 @@ func RunCmds(cmds ...[]string) error {
 	return nil
 }
 
-var (
-	parallelJobsLock      sync.Mutex
-	parallelJobsSemaphore chan int
-)
-
-func parallelJobs() chan int {
-	parallelJobsLock.Lock()
-	defer parallelJobsLock.Unlock()
+var parallelJobsSemaphore chan struct{}
 
+// parallelJobs returns a semaphore channel to limit the number of parallel jobs.
+func parallelJobs() chan struct{} {
 	if parallelJobsSemaphore == nil {
 		max := numParallel()
-		parallelJobsSemaphore = make(chan int, max)
-		log.Println("Max parallel jobs =", max)
+		parallelJobsSemaphore = make(chan struct{}, max)
+		fmt.Printf("Max parallel jobs: %d\n", max)
 	}
 
 	return parallelJobsSemaphore
 }
 
+// numParallel determines the maximum number of parallel jobs to run.
+// It considers the MAX_PARALLEL environment variable, the number of CPUs on the host,
+// and the number of CPUs reported by Docker.
 func numParallel() int {
-	if maxParallel := os.Getenv("MAX_PARALLEL"); maxParallel != "" {
-		if num, err := strconv.Atoi(maxParallel); err == nil && num > 0 {
-			return num
-		}
+	if maxParallel, err := strconv.Atoi(os.Getenv("MAX_PARALLEL")); err == nil && maxParallel > 0 {
+		return maxParallel
 	}
 
-	// To be conservative use the minimum of the number of CPUs between the host
-	// and the Docker host.
 	maxParallel := runtime.NumCPU()
 
+	// Adjust based on Docker-reported CPUs if available
 	info, err := GetDockerInfo()
 	// Check that info.NCPU != 0 since docker info doesn't return with an
-	// error status if communcation with the daemon failed.
-	if err == nil && info.NCPU != 0 && info.NCPU < maxParallel {
-		maxParallel = info.NCPU
+	// error status if communication with the daemon fails.
+	if err == nil && info.NCPU != 0 {
+		maxParallel = int(math.Min(float64(maxParallel), float64(info.NCPU)))
 	}
 
+	// Parallelize conservatively to avoid overloading the host.
+	if maxParallel >= 2 {
+		return maxParallel / 2
+	}
```

Comment:
This needs to be discussed. I did this change i.e., n/2 (compared to ~n before) because there were cases where BK agent was lost. From the Slack threads and BK notice, seems like it is happening a lot and maybe we should keep this change.

Now it is working with the change. Cross-compilations are quite expensive I think we shouldn't spawn all of them at once.

Reply:
I have a suspicion the BK agents being lost is related to a more infra level problem, I recall seeing some communication about this.

IMO we should use as much CPU as possible to ensure our builds are as fast as possible. The Buildkite agent's only job is to compile and test things quickly.

Reply:
I agree with your point. However, excessive parallelization can be problematic in certain scenarios. Prior to implementing this change, I frequently encountered the "Agent Lost" error. But after the change, none! So, I thought let me keep the change and discuss once.

Certainly, I can remove this change for now and revisit it later. The primary focus of this PR is to discuss the persistence of the build cache.